hw/xbox/xid-sdl.c

```diff
@@ -464,6 +462,61 @@ static void usb_xid_class_initfn(ObjectClass *klass, void *data)
     uc->handle_attach  = usb_desc_attach;
 }
 
+static HapticEmulationMode pick_haptic_emulation_mode(SDL_Haptic *haptic) {
+    if (SDL_HapticQuery(haptic) & SDL_HAPTIC_LEFTRIGHT) {
+        return EMU_HAPTIC_LEFT_RIGHT;
+    }
+
+    return EMU_UNSUPPORTED;
+}
+
+static int haptic_init(USBXIDState *s) {
+    if (SDL_HapticOpened(s->device_index)) {
+        fprintf(stderr, "Joystick haptic effect already in use, disabling force feedback\n");
+        return -1;
+    }
+
+    SDL_Joystick *joystick = SDL_GameControllerGetJoystick(s->sdl_gamepad);
+    s->sdl_haptic = SDL_HapticOpenFromJoystick(joystick);
+    if (s->sdl_haptic == NULL) {
+        fprintf(stderr, "Joystick doesn't support haptic effects\n");
+        return -1;
+    }
+
+    s->haptic_emulation_mode = pick_haptic_emulation_mode(s->sdl_haptic);
+
+    if (s->haptic_emulation_mode == EMU_UNSUPPORTED) {
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        fprintf(stderr, "Joystick doesn't support necissary haptic effects\n");
+        return -1;
+    }
+
+    SDL_HapticEffect effect;
+    haptic_effect_init(&effect, s->haptic_emulation_mode, 0, 0);
+
+    int effect_id = SDL_HapticNewEffect(s->sdl_haptic, &effect);
+    if (effect_id == -1) {
```

Comment:
According to [the spec](https://wiki.libsdl.org/SDL_HapticNewEffect), `SDL_HapticNewEffect` *"Returns the ID of the effect on success or a negative error code on failure."* This condition should be changed to `if (effect_id < 0)`.

Reply:
nice catch :+1: